File: tree/min_depth_of_binary_tree.py
# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param A : root node of tree
    # @return an integer
    def minDepth(self, A):

        if A is None:
            return None

        l_depth, r_depth = None, None
        if A.left is not None:
            l_depth = self.minDepth(A.left)
        if A.right is not None:
            r_depth = self.minDepth(A.right)

        if l_depth is not None and r_depth is not None:
            return (1+l_depth) if l_depth < r_depth else (1+r_depth)
        elif l_depth is not None:
            return (1+l_depth)
        elif r_depth is not None:
            return (1+r_depth)
        else:
            return 1

File: tree/min_depth_of_binary_tree.java
/**
 * Author :: @darkdefender27
 *
 * Definition for binary tree
 * class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
	public int minDepth(TreeNode a) {

	    int minDepth = 0;

	    if(a != null) {

	        Queue<TreeNode> q = new LinkedList<TreeNode>();
    	    q.add(a);

    	    int levelSize = 1;
    	    boolean foundLeafNode = false;

    	    while(!q.isEmpty() && !foundLeafNode) {

    	        minDepth++;

    	        int ct = levelSize;
    	        levelSize = 0;
    	        while(ct>0) {
    	            TreeNode el = q.remove();

        	        if(el.left == null && el.right == null) {
        	            foundLeafNode = true;
        	            break;
        	        }
        	        if(el.left != null) {
        	            q.add(el.left);
        	            levelSize++;
        	        }
        	        if(el.right != null) {
        	            q.add(el.right);
        	            levelSize++;
        	        }
        	        ct--;
    	        }
    	    }
	    }

	    return minDepth;
	}
}
